### ISSUE TITLE
New R-hat, ESS and quantile-MCSE functions

### ISSUE THREAD

**Issue by avehtari** [stan-dev/rstan#617]
#### Summary:

Add new R-hat, ESS and quantile-MCSE described in 

- Aki Vehtari, Andrew Gelman, Daniel Simpson, Bob Carpenter, Paul-Christian Bürkner (2019): Rank-normalization, folding, and localization: An improved R-hat for assessing convergence of MCMC. arXiv preprint [arXiv:1903.08008](http://arxiv.org/abs/1903.08008).

#### Description:

The reference implementation is available in [monitornew.R](https://github.com/avehtari/rhat_ess/blob/master/code/monitornew.R)

- There is a new version of `monitor()`, but for backward compatibility it would be better to use different name.
- `monitor_extra()` is needed only for the supporting experiments and should not to be implemented in rstan.
- A bit more modularity in the code would be good, so that there would be separate functions for
  - R-hat
  - bulk-ESS
  - tail-ESS
  - quantile-ESS
  - mean-ESS
  - sd-ESS
  - quantile-MCSE
  - mean-MCSE
  - sd-MCSE

**Comment by paul-buerkner** [comment]
I will start working on the PR next week.

**Comment by bgoodri** [comment]
Can we squeeze it in this weekend? Some of the functions are already in https://github.com/stan-dev/rstan/blob/develop/rstan/rstan/R/monitor.R . I am not too worried about (lack of) backward compatibility with the `monitor` function because nothing is downstream of it (unless some packages are doing something really weird). If the new stuff is strictly better than the old stuff, then I don't mind if it looks different or shows the user different things.

**Comment by avehtari** [comment]
I can help to get this in. I'd like some help with the function names and what other things should be taken into account to follow rstan and other good practices.

**Comment by paul-buerkner** [comment]
I am actually working on this right now. Expect to see a PR from me tomorrow.

**Comment by paul-buerkner** [comment]
Based on @bgoodri's comment, am I right to assume that I can just replace the old version of monitor with a new one without having to worry about any backwards compatibility?

EDIT: I just realize that monitor is actually not used in the summary output of rstan. Isn't that what we want to change as well as some point?

**Comment by avehtari** [comment]
I was just going to write that the computation in chains.hpp used by summary has not been changed yet. The plan is to change it, but I also planned to do that change at the same time for CmdStan and PyStan, but it's not possible to get that in CmdStan 2.19 anymore. The new monitor uses different acronym for the effective sample size, but Rhat is used both in old and new, but they are using different computation. I might have time to update chains.hpp in rstan if you think it's a good idea.

**Comment by betanalpha** [comment]
Changes that propagate across the interfaces, and into the core C++,
definitely need some discussion before consideration as PRs.  This
goes triply so for changes to the UX such as the summary output.

Before going too far I think that it’s a good idea to start a dev Discourse 
thread with an outline of the proposed changes and roadmap so that 
we can make a decision about how to move forwards in a more
principled way.

**Comment by bob-carpenter** [comment]
I agree with Michael that this is a big change that should be discussed
with a broader audience.

I would also be much happier seeing this implemented once in C++ in
our math lib and then reused by all the interfaces.

I realize that requires dealing with C++ and Rcpp/Cython, but it's
the only way I think we're going to keep our interfaces in synch.

It sounds like rstanarm and brms already got rid of reporting
estimates of posterior means and MCMC std error in favor of
posterior medians and MAD.

- Bob

**Comment by betanalpha** [comment]
+1 million to focusing future efforts on implementing common
functionality in the C++ before worrying about the interfaces,
and then using that common functionality to facilitate consistency
across the interfaces.

**Comment by seantalts** [comment]
Want to start a thread and outline what you want the discussion to look like? Perhaps list some desiderata or ways to evaluate such a change?

**Comment by avehtari** [comment]
> Changes that propagate across the interfaces, and into the core C++, definitely need some discussion before consideration as PRs.  This goes triply so for changes to the UX such as the summary output.

I agree completely, and that has been the plan. I did not mean this issue to lead any changes in the default output. I thought it would be useful to add non-default functions to get more experience with new options, so that it would be easier to make decisions (by whoever eventually making the decision) about the default outputs which should not be changed lightly or often.

> Before going too far I think that it’s a good idea to start a dev Discourse thread with an outline of the proposed changes and roadmap so that we can make a decision about how to move forwards in a more principled way.

I agree.

>I would also be much happier seeing this implemented once in C++ in our math lib and then reused by all the interfaces.

Note that currently Rhat and effective sample size computations are not implemented once in C++, but there are three separate C++ implementations for CmdStan, RStan and PyStan. Just adding the computations mentioned in this issue is not a big task, but I don't know how to make them once.

> +1 million to focusing future efforts on implementing common functionality in the C++ before worrying about the interfaces, and then using that common functionality to facilitate consistency across the interfaces.

I assume we need to discuss what services the interfaces need? 

I guess, we should continue the discussion in Discourse.

**Comment by bgoodri** [comment]
If you just want something to play around with, then merge it into RStan
but don't expose it. But my sense from the paper is that you all had
concluded the existing output could be wrong, so it seems strange to say we
are going to keep having potentially wrong output for 2.19.

**Comment by betanalpha** [comment]
> I assume we need to discuss what services the interfaces need?
> 
Sure.  We should really start with piping the effective sample size
calculation that was recently implemented as stand alone functions
in the C++ through to the interfaces.
> I guess, we should continue the discussion in Discourse.
> 

That would be ideal

**Comment by bob-carpenter** [comment]
> ... my sense from the paper is that you all had
> concluded the existing output could be wrong, so it seems strange to say we
> are going to keep having potentially wrong output for 2.19.

What existing output?  Is the error in the code or in the statistical methodology?

**Comment by bgoodri** [comment]
Methodology

**Comment by avehtari** [comment]
I just opened discussion at [Stan Discourse](https://discourse.mc-stan.org/t/new-r-hat-and-ess/8165)

**Comment by paul-buerkner** [comment]
I added the helper functions @avehtari suggested and now also use them in `monitor`.
